feat: new token selection modal

0xAlunara opened the pull request:
Various:
* Fixes incorrect border colors for clickable chips
* Implements correct colors for dividers
* Removes outlines on focused lists (was a black border around the entire list)
* Fixes card headers being vertically aligned to the bottom

Main:
* Adds new token select feature
* Adds stories for the new token select feature
* Removes old token select code and replaces it with the new one

Highlights:
* User search is debounced for performance
* Tried Tanstack Virtual, but had to abandon due to glitchy behaviour and even worse performance than native. The good thing is, the token icons are loaded lazily even in native React, which means they won't be loaded until the users scrolls it into view
* Adds support for token favorites. A good example use case is when creating a pool and selected a token, two tokens can be considered favorite.
* Token list is sorted first by user dollar balance, putting anything with a dollar value on top. If one has no usd value, it switches to ordering by token balance. Lastly, if that also doesn't work, it switches to ordering by token symbol
* I've added the 'hide small pools' settings to the user profile, because it makes sense to me that it's a global setting across different pages.
* The token selector can be disabled
* Specific tokens in the list can be disabled (perhaps due to blacklisting because of a hack)
* You can pass a custom settings component to the selector in case you want to add extra settings between the list and the favorite chips yourself.

I've refrained from refactoring old code as much as possible, which means that the initial fetching of the token list is still slow as hell. I don't want to inflate the PR with even more code changes than it already has.

Commit:
feat: add token select feature

Commit:
styling: add missing reactive border color to clickable chips

Commit:
styling: change color of divider according to design system

Commit:
refactor: use token select feature

Commit:
styling: remove focus outline on lists

Commit:
styling: fix card header being aligned to bottom

Commit:
styling: remove unnecessary checkbox horizontal aligment to end

Commit:
styling: fix misaligned modal close button icon

Commit:
fix: recursive url query toggling by removng hidesmallpools from search params

Commit:
fix: create pool token list lost its sorting by volume

Commit:
feat: add volume sorting to token selector

Commit:
merge: main

Commit:
fix: incorrect export

0xAlunara committed:
Merge branch 'main' into feat/improve-token-selection-modal

DanielSchiavini commented on packages/curve-ui-kit/src/features/select-token/ui/modal/TokenList.tsx:
```diff
@@ -0,0 +1,163 @@
+import { FavoriteTokens } from './FavoriteTokens'
+import { TokenOption } from './TokenOption'
+
+// Prevent all the token options from re-rendering if only the balance of a single one has changed.
```
This shouldn't be necessary, it should happen automatically :thinking:

0xAlunara replied:
I tried with and without and didn't notice / measure any difference, so I've removed it.

0xAlunara replied:
I really need to give virtualization a second try though, when you first open the modal with all tokens the performance is dogshit imo. Not the case when I limit the list to just 5 items. It's also not the re-renders causing the performance hog, checked it with a `console.log(symbol)` inside `TokenOption.tsx`.

DanielSchiavini replied:
The comment is still left. Did you check the performance profiler?

0xAlunara replied:
I've checked the profiler, and could not find any meaningful slowdown in the token list itself. The performance is not bad at all in production, so I chalk it up to a slow laptop on top of local dev mode. It seems the headers and footers with their icons are bigger performance hogs as well. On top of that, after discussion with tutti we're no longer showing the entire token list at once, but will have a "show more" button, as per the new Figma design. Will resolve this for now.

DanielSchiavini commented on packages/curve-ui-kit/src/features/select-token/ui/modal/TokenOption.tsx:
```diff
@@ -0,0 +1,61 @@
+        }),
+      }}
+    >
+      {/* I'm abusing the symbol here to deliberately show a tooltip for the address instead */}
```
why? :thinking:

0xAlunara replied:
I've experimented with showing the address on hover and keyboard focus at the place where the label would be, but because we're not using labels it causes the token symbol to move up. See video attached, the symbol typography has to move up to make room for the address to show up. When no longer hovering or focusing the item the address would go away and the symbol would move down again.

This looks very jumpy and distracting so I've opted to keep the token on hover for now

https://github.com/user-attachments/assets/5233f53e-e299-4904-b3e5-6764be2b39cd

0xAlunara replied:
Renamed the property from 'symbol' to 'tooltip' to more clearly indicate its intent.
As discussed with tutti, we've decided to implement two sections in the modal: "My tokens" and "All tokens"

In "My tokens", we will show balances as we currently do. In "All tokens" we will show the token address (where currently nothing is shown).

DanielSchiavini commented on apps/main/src/dex/components/PageCreatePool/TokensInPool/SelectTokenButton.tsx:
```diff
@@ -0,0 +1,289 @@
+  onSelectionChange: (selectedAddress: React.Key) => void
+}
+
+const ComboBoxTokenPicker = ({
```
Was this file moved? What was the name? Diff isn't showing.

0xAlunara replied:
Lol seems to have been a copy-paste error from before my time. I'll fix it though

DanielSchiavini commented on apps/main/src/dex/components/PagePool/Swap/index.tsx:
```diff
@@ -88,15 +86,32 @@ const Swap = ({
+        .filter((token) => !!token)
+        .map((token) => ({
+          chain: network?.networkId ?? '',
+          address: token?.address as `0x${string}`,
```
```suggestion
          address: token?.address as Address,
```

DanielSchiavini commented on apps/main/src/dex/components/PagePool/Swap/index.tsx:
```diff
@@ -88,15 +86,32 @@ const Swap = ({
+          chain: network?.networkId ?? '',
+          address: token?.address as `0x${string}`,
+          symbol: token?.symbol,
+          label: '',
```
Why is this empty? And why do we use an empty string instead of undefined?
```suggestion
```

0xAlunara replied:
I prefer explicitness and want to avoid null & undefined. Will make this one optional though

DanielSchiavini commented on apps/main/src/dex/components/PagePool/Swap/index.tsx:
```diff
@@ -88,15 +86,32 @@ const Swap = ({
+      selectList: selectList
+        .filter((token) => !!token)
+        .map((token) => ({
+          chain: network?.networkId ?? '',
```
if chain is optional it should be optional not an empty string

0xAlunara replied:
TokenIcon chain used be required, but I've now made it optional and default to `''` inside the component itself

DanielSchiavini commented on apps/main/src/dex/components/PageRouterSwap/index.tsx:
```diff
@@ -98,30 +97,54 @@ const QuickSwap = ({
-    () => getTokensObjList(selectFromList ?? selectToList, tokensMapper),
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-    [selectFromList, selectToListStr, tokensMapperStr],
+  function getTokensObjList(tokensList: string[] | undefined, tokensMapper: TokensMapper | undefined) {
```
this should be in the global scope or be included in the `useMemo` dependencies, not sure why we don't get a warn

DanielSchiavini commented on apps/main/src/dex/components/PageRouterSwap/index.tsx:
```diff
@@ -98,30 +97,54 @@ const QuickSwap = ({
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-    [selectFromList, selectToListStr, tokensMapperStr],
+  function getTokensObjList(tokensList: string[] | undefined, tokensMapper: TokensMapper | undefined) {
+    if (!tokensList || tokensList.length === 0 || !tokensMapper || Object.keys(tokensMapper).length === 0) return []
```
```suggestion
    // import { isEmpty } from 'lodash'
    if (isEmpty(tokensList) || isEmpty(tokensMapper)) return []
```

0xAlunara replied:
Had to add an exclamation mark to `return tokensList!.map()` as isEmpty works with `undefined`, but doesn't alter the type of `tokensList` from `string[] | undefined` to `string[]` even though we know for sure it's not undefined.

DanielSchiavini commented on apps/main/src/dex/components/PageRouterSwap/index.tsx:
```diff
@@ -98,30 +97,54 @@ const QuickSwap = ({
+      getTokensObjList(selectFromList ?? selectToList, tokensMapper)
+        .filter((token) => !!token)
+        .map((token) => ({
+          chain: network?.networkId ?? '',
+          address: token?.address as `0x${string}`,
+          symbol: token?.symbol,
+          label: '',
+          volume: token?.volume ?? 0,
```
Zero and empty strings should not be the fallback

```suggestion
          chain: network?.networkId,
          address: token?.address as Address,
          symbol: token?.symbol,
          volume: token?.volume,
```

DanielSchiavini commented on apps/main/src/dex/components/PageRouterSwap/index.tsx:
```diff
@@ -98,30 +97,54 @@ const QuickSwap = ({
+  const tokensFrom = useMemo(
+    () =>
+      getTokensObjList(selectFromList ?? selectToList, tokensMapper)
+        .filter((token) => !!token)
+        .map((token) => ({
```
Why isn't `filter` and `map` part of `getTokensObjList` as it's repeated

DanielSchiavini commented on apps/main/src/dex/features/add-gauge-reward-token/ui/TokenSelector.tsx:
```diff
@@ -24,23 +25,34 @@ export const TokenSelector = ({
+          ),
+      )
+      .map((token) => ({
+        chain: network?.networkId ?? '',
```
this is repeated code

0xAlunara replied:
My bad, there were multiple instances so I lost track.

For the mapping of 'old' token objects to the TokenOption type required by the token selector feature I've written a small utility function that, given a network, converts it into such:
```ts
export const toTokenOption =
  (chain?: string) =>
  (token: Token): TokenOption => ({
    chain,
    address: token.address as Address,
    symbol: token.symbol,
    volume: token.volume,
  })
```
Example usage:
```ts
tokens.map(toTokenOption(network?.networkId))
```

The `chain`, `label` and `volume` properties in the token option type itself have also been made optional:
```ts
export type TokenOption = {
  address: Address
  symbol: string
  chain?: string
  label?: string
  volume?: number
}
```

DanielSchiavini commented on apps/main/src/dex/features/deposit-gauge-reward/ui/AmountTokenInput.tsx:
```diff
@@ -54,30 +53,40 @@ export const AmountTokenInput = ({ chainId, poolId }: { chainId: ChainId; poolId
+          activeRewardTokens.some((rewardToken) => isAddressEqual(rewardToken as Address, token.address as Address)),
+      )
+      .map((token) => ({
+        chain: networkId ?? '',
```
repeated code

DanielSchiavini commented on apps/main/src/dex/features/deposit-gauge-reward/ui/AmountTokenInput.tsx:
```diff
@@ -54,30 +53,40 @@ export const AmountTokenInput = ({ chainId, poolId }: { chainId: ChainId; poolId
+    const filteredTokens = Object.values(tokensMapper)
+      .filter(
+        (token): token is Token =>
+          token !== undefined &&
```
What if it's null?
```suggestion
          token &&
```

0xAlunara replied:
That doesn't work, but `!!token` does

DanielSchiavini commented on apps/main/src/dex/features/deposit-gauge-reward/ui/HelperFields.tsx:
```diff
@@ -1,18 +1,25 @@
-    data: [tokenUsdRate],
-  } = useTokensUSDRates([rewardTokenId])
+  const usdRatesMapper = useStore((state) => state.usdRates.usdRatesMapper)
+  const tokens = [rewardTokenId]
+  const tokensKey = JSON.stringify(tokens)
+
+  const [tokenUsdRate] = useMemo(
+    () => tokens.map((token) => (token ? usdRatesMapper[token] : undefined)),
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [tokensKey, usdRatesMapper],
+  )
```
Why do we need this complicated JSON stringify and ignoring eslint rules?

```suggestion
  const tokenUsdRate = rewardTokenId && usdRatesMapper[rewardTokenId]
```

0xAlunara replied:
Refactoring this away with a tanstack query in a future PR

0xAlunara replied:
It's not new, I just inlined whatever `useTokensUSDRates` used to be as this was the only place it was called. As said, will be a tanstack query in the future.

DanielSchiavini commented on apps/main/src/dex/store/createQuickSwapSlice.ts:
```diff
@@ -410,11 +409,9 @@ const createQuickSwapSlice = (set: SetState<State>, get: GetState<State>): Quick
-        ['desc'],
-      ).map(({ address }) => address)
+      const selectToList = Object.entries(tokensMapper)
+        .map(([_, v]) => v!)
```
This looks incorrect - if `tokensMapper` has no `null`/`undefined` why is it typed so?

0xAlunara replied:
Probably because `export type TokensMapper = { [tokenAddress: string]: Token | undefined }`

If I remove the exclamation mark I actually get a type error that the `address` in `.map(({ address }) => address)` might be `undefined` as the type is `Token | undefined`. It's a bit silly because then I don't even know if you can have a key whose value is undefined?

0xAlunara replied:
I feel like we're working around a typescript quirk here, but nonetheless I've added a `.filter((token) => !!token)`

DanielSchiavini commented on packages/curve-ui-kit/src/features/select-token/ui/TokenSelectButton.tsx:
```diff
@@ -0,0 +1,95 @@
+  <Stack direction="row" gap={Spacing.xxs} alignItems="center">
+    <TokenIcon
+      blockchainId={token.chain}
+      symbol="" // No need for tooltip of symbol as it's displayed right next to it
```
```suggestion
      // No need for tooltip of symbol as it's displayed right next to it
```

0xAlunara replied:
Had to make symbol an optional property that defaults to the empty string for this

DanielSchiavini commented on packages/curve-ui-kit/src/features/user-profile/settings/hide-small-pools/HideSmallPools.tsx:
```diff
@@ -0,0 +1,22 @@
+import { SettingLabel, Setting } from '../Setting'
+import useUserProfileStore from '../../store'
+
+export const HideSmallPools = () => {
```
```suggestion
export const HideSmallPoolsSetting = () => {
```

DanielSchiavini commented on packages/curve-ui-kit/src/themes/components.ts:
```diff
@@ -97,6 +104,15 @@ export const createComponents = (design: DesignSystem, typography: TypographyOpt
+    styleOverrides: {
+      root: {
+        '&:focus': {
+          outline: 'none',
```
Are you sure we want to disable this? It might be useful for accessibility. How about `focus-visible`?
For buttons and icon buttons I use `'&:focus-visible': { borderColor: Focus_Outline }` as discussed with UX.

0xAlunara replied:
The menu list has an auto-focus so you can use keyboard navigation immediately to navigate the list. However, that means it immediately renders with a damn awful black border. I could change `&:focus` to `&:focus-visible`, but showing any kind of outline is not an option, so neither would be using `Focus_Outline`

![image](https://github.com/user-attachments/assets/4981a5c6-3b34-4251-8959-2be40d564f7f)

Commit:
chore: rename component filename

Commit:
refactor: token icon symbol optional (defaults to empty string)

0xAlunara commented on packages/curve-ui-kit/src/features/select-token/ui/modal/TokenList.tsx:
```diff
@@ -0,0 +1,163 @@
+          // If USD balances are equal, compare raw balances
+          if (aBalance !== bBalance) return bBalance - aBalance
+
+          // Sort by volume in descending order
```
It matches the original ordering though, so at least it's not different from what people are used to 😅

Commit:
styling: error in token modal alert grows in height

Commit:
fix: onSearch would be invoked even if search didn't change

Commit:
fix: error keeps being visible after search has changed

Commit:
fix: error would be visible even after modal closing

0xAlunara commented on packages/curve-ui-kit/src/features/select-token/ui/modal/TokenList.tsx:
```diff
@@ -0,0 +1,163 @@
+      )}
+
+      {error ? (
+        <Alert variant="filled" severity="error">
```
I've made the error message grow in height as needed. The height is actually fixed to `IconSize.sm` in `defineMuiAlertTitle`. I didn't really want to introduce an entire variant for this one off, so I've opted for `sx={{ '&': { height: 'auto' } }}` for now.

The error will also go away when you close the modal or change the search value.

Not sure what's the best way to tack on a friendly error message. I could change any error to "invalid token" but maybe it's not related to the token but the RPC. Or maybe there's something else, and then you'd want to know the full error. Not sure we can just look at the error message and search for "Multicall3", replace it with "Invalid token" and call it a day

DanielSchiavini replied:
I think we should only have a generic error as title, the details can be in the alert body

0xAlunara replied:
You're right, I was so focused on the title I forgot you can use an actual body message.

I've added the copy button (smallest sizen, ghost) in an unconventional way, because the button would take up an entire column, leaving a lot of unused whitespace that the message could've used. So for now I've manually experimented with some absolute positioning. This is a one off, and if we want to make this the default behavior I'd rather implement it properly in a different PR.

I've also added word wrapping to the message. I have a hunch not all users actually use the copy function and send a screenshot instead. This would make the error more readable as well; scrollbars would be annoying in this case.

![image](https://github.com/user-attachments/assets/39876f70-cee8-49d7-b0e5-7a2fd34b8d61)

Commit:
refactor: remove useless memo

Commit:
chore: fix incorrect variable name

Commit:
refactor: use Address type

Commit:
refactor: make token option label optional

Commit:
refactor: make token option volume optional

Commit:
refactor: convert undefined check to truthy check

Commit:
refactor: TokenIcon chain id is now optional

Commit:
refactor: use lodash isEmpty

Commit:
refactor: move getTokensObjList to global scope

Commit:
refactor: deduplicate filter and map for getTokensObjList

Commit:
refactor: rely on default value

Commit:
refactor: add toTokenOption util function

Commit:
styling: token list error is now a message with a copy button

Commit:
refactor: rename 'symbol' to 'tooltip' for TokenIcon

Commit:
refactor: remove unused import

Commit:
refactor: replace null assertion with truthy filter

Commit:
refactor: don't make needless copy

Commit:
refactor: re-use existing search functionality for tokens

0xAlunara commented on packages/curve-ui-kit/src/features/select-token/ui/modal/TokenList.tsx:
```diff
@@ -0,0 +1,163 @@
+
+  // List of all token options
+  const options = useMemo(() => {
+    const tokensFiltered = [...tokens].filter((token) =>
```
Figured I'd keep it more simple, but yeah, maybe common behavior across the board is more desired.

Commit:
fix: whitespace shouldn't be a valid search value

Commit:
refactor: make error optional

0xAlunara commented on apps/main/src/dex/components/PageCreatePool/TokensInPool/SelectTokenButton.tsx:
```diff
@@ -0,0 +1,289 @@
+  const { basePools, basePoolsLoading } = useStore((state) => state.pools)
+  const { swapType } = useStore((state) => state.createPool)
+
+  const [error, setError] = useState('')
```
Copied it from somewhere else in the code, but sure, made it optional

Commit:
refactor: smoller code

Commit:
feat: show token address if no balance

Commit:
feat: remove token address tooltip on token icon hover

Commit:
styling: add default typography color transition

Commit:
styling: use inverted colors for token select option hover

Commit:
refactor: remove token settings page but keep wiring for future manage token list page

DanielSchiavini committed:
Merge branch 'main' of github.com:curvefi/curve-frontend into feat/improve-token-selection-modal-merge

Commit:
refactor: extract search input field into its own component with internal debounced state

Commit:
refactor: extract error alert into its own component

Commit:
feat: introduce 'my tokens' and 'all tokens' sections

Commit:
fix: incorrect variable name

Commit:
styling: fix small card header height changing height

Commit:
feat: add 'show more' buttons

Commit:
feat: don't show 'all tokens' section title if no 'my tokens'

Commit:
styling: fix weird ass chip icon margins and paddings

Commit:
styling: add expand more icon to show more button

Commit:
feat: no longer hide small pools on swap page

Commit:
refactor: remove unused memo

Commit:
refactor: simplify token list generation for quickswap

Commit:
fix: modals steal focus from elements inside that have autoFocus

Commit:
fix: a few tab issues

Commit:
styling: give chips the same hover effect when they're focused

Commit:
fix: invert theme when selecting token through keyboard navigation

Commit:
fix: use i18n

Commit:
fix: don't search by non-existant key

Commit:
chore: add comment

Commit:
refactor: dont re-render when isOpen changes, mui is smart enough to not render non-open dialogs

0xAlunara commented:
Major changes since last version:

* Tokens are now split up in two sections called "My tokens" and "All tokens" if applicable.
* 'My tokens' shows 5 tokens, 'All tokens' shows 10 tokens. If there's more you can click the 'Show more' buttons.
* Tokens without a balance now show the shortened token address for verification (the idea is, if you don't own it you prob want to verify it's the correct one. And if you do own it you prob don't care about the address anymore as you already trust it)
* Chips with icons have their padding and margin fixed
* Lots of speed optimizations. The biggest one is that `createQuickSwapSlice` no longer has two different 'from' and 'to' lists, and the sorting is gone as it's now done internally by the new select token modal.
* Hover and focus-visible now use inverted themes. Not ideal but I got really tired so I'm calling it a day. Improvements regarding hover and focus states for MUI's MenuItem are for another PR.
* Errors are rendered more nicely.
* A few focus / keyboard navigation related fixed.
* All review feedback has been processed.

Commit:
fix: dont show divider for section title if there's none

0xAlunara commented on apps/main/src/dex/components/PageCreatePool/TokensInPool/SelectTokenButton.tsx:
```diff
@@ -0,0 +1,289 @@
+  onSelectionChange: (selectedAddress: React.Key) => void
+}
+
+const SelectTokenButton = ({
```
It's indeed the old file, but moved and trimmed down. A lot of code has been deleted now we have `TokenSelectorModal`, but not everything for the token selection page when creating a pool. So what you see is what remained and could not be replaced by the new `TokenSelectorModal` in its entirety.

0xAlunara commented on apps/main/src/dex/features/deposit-gauge-reward/ui/AmountTokenInput.tsx:
```diff
@@ -54,30 +54,34 @@ export const AmountTokenInput = ({ chainId, poolId }: { chainId: ChainId; poolId
+    const filteredTokens = Object.values(tokensMapper)
+      .filter(
+        (token): token is Token =>
+          !!token &&
```
Doesn't work:
```
Type 'boolean | undefined' is not assignable to type 'boolean'.
  Type 'undefined' is not assignable to type 'boolean'.
```

0xAlunara commented on packages/curve-ui-kit/src/features/select-token/ui/modal/SearchInput.tsx:
```diff
@@ -0,0 +1,38 @@
+  const [search, setSearch] = useState('')
+  const lastSearch = useRef(search)
+
+  useEffect(() => {
```
Been going back and forth of this as I came across multiple debounce implemenations. So naturally, I created another version :D
Will create a small PR

0xAlunara commented on packages/curve-ui-kit/src/features/select-token/ui/modal/TokenList.tsx:
```diff
@@ -0,0 +1,236 @@
+            endIcon={<ExpandMoreIcon />}
+            onClick={onShowAll}
+            // Override variant button height to match menu list item height
+            sx={{ '&': { height: `${ButtonSize.md} !important` } }}
```
Ah right, forgot to remove it after '&' didn't do the trick.

DanielSchiavini commented on apps/main/src/dex/components/PageCreatePool/TokensInPool/SelectTokenButton.tsx:
```diff
@@ -0,0 +1,289 @@
+          updateUserAddedTokens(filterValueLowerCase, token[0].symbol, false, isBasePool)
+        } catch (error) {
+          console.log(error)
+          setError(error)
```
@0xAlunara error is expected to be string!

0xAlunara replied:
I think there should be a lint rule now that warns you when you're not typechecking `error` in a `catch` clause 🤔

DanielSchiavini replied:
yeah that's how I found it in #2611 :nerd_face:
